drv/i2c-devices/src/lm5066.rs

```diff
@@ -0,0 +1,233 @@
+use pmbus::commands::*;
+use ringbuf::*;
+use userlib::units::*;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Error {
+    BadRead { cmd: u8, code: ResponseCode },
+    BadWrite { cmd: u8, code: ResponseCode },
+    BadData { cmd: u8 },
+    BadValidation { cmd: u8, code: ResponseCode },
+    InvalidData { err: pmbus::Error },
+    InvalidDeviceSetup,
+}
+
+impl From<BadValidation> for Error {
+    fn from(value: BadValidation) -> Self {
+        Self::BadValidation {
+            cmd: value.cmd,
+            code: value.code,
+        }
+    }
+}
+
+impl From<pmbus::Error> for Error {
+    fn from(err: pmbus::Error) -> Self {
+        Error::InvalidData { err }
+    }
+}
+
+impl From<Error> for ResponseCode {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::BadRead { code, .. } => code,
+            Error::BadWrite { code, .. } => code,
+            Error::BadValidation { code, .. } => code,
+            _ => panic!(),
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+#[allow(dead_code)]
+struct Coefficients {
+    current: pmbus::Coefficients,
+    power: pmbus::Coefficients,
+}
+
+#[derive(Copy, Clone)]
+pub enum CurrentLimitStrap {
+    /// CL pin is strapped to VDD, denoting the Low setting
+    VDD,
+    /// CL pin is strapped to GND (or left floating), denoting the high setting
+    GND,
+}
+
+pub struct Lm5066 {
+    /// Underlying I2C device
+    device: I2cDevice,
+    /// Value of the rsense resistor, in milliohms
+    rsense: i32,
+    /// Sense of current limit pin
+    cl: CurrentLimitStrap,
+    /// Our (cached) coefficients
+    coefficients: Cell<Option<Coefficients>>,
+    /// Our (cached) device setup
+    device_setup: Cell<Option<lm5066::DEVICE_SETUP::CommandData>>,
+}
+
+impl core::fmt::Display for Lm5066 {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "lm5066: {}", &self.device)
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    CurrentCoefficients(pmbus::Coefficients),
+    PowerCoefficients(pmbus::Coefficients),
+    DeviceSetup(lm5066::DEVICE_SETUP::CommandData),
+    None,
+}
+
+ringbuf!(Trace, 8, Trace::None);
+
+impl Lm5066 {
+    pub fn new(
+        device: &I2cDevice,
+        rsense: Ohms,
+        cl: CurrentLimitStrap,
+    ) -> Self {
+        Self {
+            device: *device,
+            rsense: (rsense.0 * 1000.0).round() as i32,
+            cl,
+            coefficients: Cell::new(None),
+            device_setup: Cell::new(None),
+        }
+    }
+
+    fn read_device_setup(
+        &self,
+    ) -> Result<lm5066::DEVICE_SETUP::CommandData, Error> {
+        if let Some(ref device_setup) = self.device_setup.get() {
+            return Ok(*device_setup);
+        }
+
+        let device_setup = pmbus_read!(self.device, lm5066::DEVICE_SETUP)?;
+        ringbuf_entry!(Trace::DeviceSetup(device_setup));
+        self.device_setup.set(Some(device_setup));
+
+        Ok(device_setup)
+    }
+
+    //
```

Comment:
Nit: `///` so that this is recognized as a docstring

Reply:
Does that make sense for a non-public function? (I have always used docstrings for stuff that I want to show up in rustdoc.)

Reply:
I find the distinction meaningful because my editor highlights them differently, so it's easier to spot the beginnings of functions:

![Screenshot 2025-02-04 at 4 12 55 PM](https://github.com/user-attachments/assets/78efe7b2-2c0c-4fdc-ab02-6ca40975667c)